File: app/src/main/java/com/teambrella/android/ui/home/HomeCardsFragment.java
package com.teambrella.android.ui.home;

import android.content.Context;
import android.content.res.Resources;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v4.app.Fragment;
import android.support.v4.app.FragmentStatePagerAdapter;
import android.support.v4.view.ViewPager;
import android.text.Html;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.TextView;

import com.google.gson.Gson;
import com.google.gson.GsonBuilder;
import com.google.gson.JsonArray;
import com.google.gson.JsonObject;
import com.squareup.picasso.Picasso;
import com.squareup.picasso.RequestCreator;
import com.teambrella.android.R;
import com.teambrella.android.api.TeambrellaModel;
import com.teambrella.android.api.model.json.JsonWrapper;
import com.teambrella.android.api.server.TeambrellaServer;
import com.teambrella.android.image.TeambrellaImageLoader;
import com.teambrella.android.ui.IMainDataHost;
import com.teambrella.android.ui.base.ADataFragment;
import com.teambrella.android.ui.chat.ChatActivity;
import com.teambrella.android.util.AmountCurrencyUtil;
import com.teambrella.android.util.TeambrellaDateUtils;

import io.reactivex.Notification;
import jp.wasabeef.picasso.transformations.CropCircleTransformation;
import jp.wasabeef.picasso.transformations.MaskTransformation;

/**
 * Home Cards Fragment.
 */
public class HomeCardsFragment extends ADataFragment<IMainDataHost> {


    private TextView mHeader;
    private ViewPager mCardsPager;
    private CardAdapter mAdapter;
    private LinearLayout mPagerIndicator;

    @Nullable
    @Override
    public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        View view = inflater.inflate(R.layout.fragment_home_cards, container, false);
        mHeader = view.findViewById(R.id.home_header);
        mCardsPager = view.findViewById(R.id.cards_pager);
        mPagerIndicator = view.findViewById(R.id.page_indicator);
        mCardsPager.setPageMargin(40);
        return view;
    }


    @Override
    public void onDestroyView() {
        super.onDestroyView();
        mAdapter = null;
    }

    @Override
    protected void onDataUpdated(Notification<JsonObject> notification) {
        if (notification.isOnNext()) {
            JsonWrapper response = new JsonWrapper(notification.getValue());
            JsonWrapper data = response.getObject(TeambrellaModel.ATTR_DATA);
            mHeader.setText(getString(R.string.welcome_user_format_string, data.getString(TeambrellaModel.ATTR_DATA_NAME).trim().split(" ")[0]));
            JsonArray cards = data.getJsonArray(TeambrellaModel.ATTR_DATA_CARDS);
            if (mAdapter == null) {
                mCardsPager.setAdapter(mAdapter = new CardAdapter(cards));
                LayoutInflater inflater = LayoutInflater.from(getContext());
                for (int i = 0; i < cards.size(); i++) {
                    View view = inflater.inflate(R.layout.home_card_pager_indicator, mPagerIndicator, false);
                    view.setSelected(mCardsPager.getCurrentItem() == i);
                    mPagerIndicator.addView(view);
                }

                mCardsPager.addOnPageChangeListener(new ViewPager.OnPageChangeListener() {
                    @Override
                    public void onPageScrolled(int position, float positionOffset, int positionOffsetPixels) {

                    }

                    @Override
                    public void onPageSelected(int position) {
                        for (int i = 0; i < mPagerIndicator.getChildCount(); i++) {
                            mPagerIndicator.getChildAt(i).setSelected(position == i);
                        }
                    }

                    @Override
                    public void onPageScrollStateChanged(int state) {
                        ((HomeFragment) getParentFragment()).setRefreshingEnable(state == ViewPager.SCROLL_STATE_IDLE);
                    }
                });
            } else {
                mAdapter.setData(cards);
            }
        }
    }


    public final class CardAdapter extends FragmentStatePagerAdapter {

        private JsonArray mCards;

        CardAdapter(JsonArray cards) {
            super(HomeCardsFragment.this.getChildFragmentManager());
            mCards = cards;
        }

        @Override
        public Fragment getItem(int position) {
            return CardsFragment.getInstance(mCards.get(position).toString(), mDataHost.getTeamId(), mDataHost.getCurrency());
        }

        @Override
        public int getCount() {
            return mCards.size();
        }

        public void setData(JsonArray cards) {
            mCards = cards;
            notifyDataSetChanged();
        }

        @Override
        public int getItemPosition(Object object) {
            return POSITION_NONE;
        }
    }


    public static final class CardsFragment extends Fragment {

        private static final String EXTRA_DATA = "data";
        private static final String EXTRA_CURRENCY = "currency";
        private static final String EXTRA_TEAM_ID = "team_id";

        private JsonWrapper mCard;
        private int mTeamId;
        private String mCurrency;

        public static CardsFragment getInstance(String data, int teamId, String currency) {
            CardsFragment fragment = new CardsFragment();
            Bundle args = new Bundle();
            args.putString(EXTRA_DATA, data);
            args.putString(EXTRA_CURRENCY, currency);
            args.putInt(EXTRA_TEAM_ID, teamId);
            fragment.setArguments(args);
            return fragment;
        }

        @Override
        public void onCreate(@Nullable Bundle savedInstanceState) {
            super.onCreate(savedInstanceState);
            Gson gson = new GsonBuilder().create();
            mCard = new JsonWrapper(gson.fromJson(getArguments().getString(EXTRA_DATA), JsonObject.class));
            mTeamId = getArguments().getInt(EXTRA_TEAM_ID);
            mCurrency = getArguments().getString(EXTRA_CURRENCY);
        }

        @Nullable
        @Override
        public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
            Picasso picasso = TeambrellaImageLoader.getInstance(getContext()).getPicasso();

            View view = inflater.inflate(R.layout.home_card_claim, container, false);

            ImageView icon = view.findViewById(R.id.icon);
            TextView message = view.findViewById(R.id.message_text);
            TextView unread = view.findViewById(R.id.unread);
            TextView amountWidget = view.findViewById(R.id.amount_widget);
            TextView teamVote = view.findViewById(R.id.team_vote);
            TextView title = view.findViewById(R.id.title);
            TextView subtitle = view.findViewById(R.id.subtitle);
            TextView leftTitle = view.findViewById(R.id.left_title);


            int itemType = mCard.getInt(TeambrellaModel.ATTR_DATA_ITEM_TYPE);


            leftTitle.setText(itemType == TeambrellaModel.FEED_ITEM_TEAMMATE ? R.string.limit : R.string.claimed);


            RequestCreator requestCreator = picasso.load(TeambrellaModel.getImage(TeambrellaServer.BASE_URL, mCard.getObject(), TeambrellaModel.ATTR_DATA_SMALL_PHOTO_OR_AVATAR));

            if (itemType == TeambrellaModel.FEED_ITEM_TEAMMATE
                    || itemType == TeambrellaModel.FEED_ITEM_TEAM_CHAT) {
                requestCreator.transform(new CropCircleTransformation());
            } else {
                Resources resources = getContext().getResources();
                requestCreator.resize(resources.getDimensionPixelSize(R.dimen.image_size_48), resources.getDimensionPixelSize(R.dimen.image_size_48))
                        .centerCrop().transform(new MaskTransformation(getContext(), R.drawable.teammate_object_mask));
            }

            requestCreator.into(icon);

            message.setText(Html.fromHtml(mCard.getString(TeambrellaModel.ATTR_DATA_TEXT)));

            int unreadCount = mCard.getInt(TeambrellaModel.ATTR_DATA_UNREAD_COUNT);

            unread.setVisibility(unreadCount > 0 ? View.VISIBLE : View.GONE);
            unread.setText(mCard.getString(TeambrellaModel.ATTR_DATA_UNREAD_COUNT));

            AmountCurrencyUtil.setAmount(amountWidget, mCard.getFloat(TeambrellaModel.ATTR_DATA_AMOUNT), getArguments().getString(EXTRA_CURRENCY));
            if (itemType == TeambrellaModel.FEED_ITEM_TEAMMATE) {
                teamVote.setText(getString(R.string.risk_format_string, mCard.getFloat(TeambrellaModel.ATTR_DATA_TEAM_VOTE)));
            } else {
                teamVote.setText(Html.fromHtml(getString(R.string.home_team_vote_format_string, Math.round(mCard.getFloat(TeambrellaModel.ATTR_DATA_TEAM_VOTE) * 100))));
            }


            switch (itemType) {
                case TeambrellaModel.FEED_ITEM_CLAIM:
                    title.setText(getString(R.string.claim_title_format_string, mCard.getInt(TeambrellaModel.ATTR_DATA_ITEM_ID)));
                    break;
                case TeambrellaModel.FEED_ITEM_TEAM_CHAT:
                    title.setText(mCard.getString(TeambrellaModel.ATTR_DATA_CHAT_TITLE));
                    break;
                case TeambrellaModel.FEED_ITEM_TEAMMATE:
                    title.setText(R.string.application);
                    break;
            }

            subtitle.setText(TeambrellaDateUtils.getDatePresentation(getContext()
                    , TeambrellaDateUtils.TEAMBRELLA_UI_DATE
                    , mCard.getString(TeambrellaModel.ATTR_DATA_ITEM_DATE)));

            view.setOnClickListener(v -> {
                Context context = getContext();
                switch (itemType) {
                    case TeambrellaModel.FEED_ITEM_CLAIM:
                        ChatActivity.startClaimChat(context
                                , mTeamId
                                , mCard.getInt(TeambrellaModel.ATTR_DATA_ITEM_ID)
                                , mCard.getString(TeambrellaModel.ATTR_DATA_MODEL_OR_NAME)
                                , TeambrellaImageLoader.getImageUri(mCard.getString(TeambrellaModel.ATTR_DATA_SMALL_PHOTO_OR_AVATAR))
                                , mCard.getString(TeambrellaModel.ATTR_DATA_TOPIC_ID)
                                , mCurrency);
                        break;
                    case TeambrellaModel.FEED_ITEM_TEAM_CHAT:
                        ChatActivity.startFeedChat(context
                                , mCard.getString(TeambrellaModel.ATTR_DATA_CHAT_TITLE)
                                , mCard.getString(TeambrellaModel.ATTR_DATA_TOPIC_ID));
                        break;
                    default:
                        ChatActivity.startTeammateChat(context, mTeamId
                                , mCard.getString(TeambrellaModel.ATTR_DATA_ITEM_USER_ID)
                                , null
                                , TeambrellaImageLoader.getImageUri(mCard.getString(TeambrellaModel.ATTR_DATA_SMALL_PHOTO_OR_AVATAR))
                                , mCard.getString(TeambrellaModel.ATTR_DATA_TOPIC_ID)
                                , mCurrency);
                        break;
                }
            });

            return view;

        }
    }

}

File: app/src/main/java/com/teambrella/android/ui/home/HomeCardsFragment.kt
package com.teambrella.android.ui.home

import android.os.Bundle
import android.text.Html
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.LinearLayout
import android.widget.TextView
import androidx.constraintlayout.widget.ConstraintLayout
import androidx.fragment.app.Fragment
import androidx.fragment.app.FragmentStatePagerAdapter
import androidx.lifecycle.Observer
import androidx.viewpager.widget.PagerAdapter
import androidx.viewpager.widget.ViewPager
import com.google.gson.JsonArray
import com.google.gson.JsonObject
import com.teambrella.android.R
import com.teambrella.android.api.*
import com.teambrella.android.ui.IMainDataHost
import com.teambrella.android.ui.base.ADataFragment
import com.teambrella.android.ui.base.createDataFragment
import com.teambrella.android.ui.chat.ChatActivity
import com.teambrella.android.ui.util.setAvatar
import com.teambrella.android.ui.util.setImage
import com.teambrella.android.ui.util.setUnreadCount
import com.teambrella.android.util.AmountCurrencyUtil
import io.reactivex.Notification
import java.util.*

class HomeCardsFragment : ADataFragment<IMainDataHost>() {

    private val header: TextView? by ViewHolder(R.id.home_header)
    private val subHeader: TextView? by ViewHolder(R.id.home_sub_header)
    private val cardsPager: ViewPager? by ViewHolder(R.id.cards_pager)
    private var pagerAdapter: CardsAdapter? = null
    private val pagerIndicator: LinearLayout? by ViewHolder(R.id.page_indicator)

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? =
            inflater.inflate(R.layout.fragment_home_cards, container, false)

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        cardsPager?.pageMargin = 20
        val adapter = CardsAdapter()
        cardsPager?.adapter = adapter
        pagerAdapter = adapter
        super.onViewCreated(view, savedInstanceState)
        dataHost.getObservable(tags[0]).observe(this, adapter)
    }

    override fun onDataUpdated(notification: Notification<JsonObject>) {
        val data = notification.takeIf { it.isOnNext }?.value?.data
        data?.let { _data ->
            header?.text = getString(R.string.welcome_user_format_string, _data.name?.trim()?.split(" ".toRegex())!![0])
            subHeader?.visibility = View.VISIBLE
            val inflater = LayoutInflater.from(context)
            pagerIndicator?.removeAllViews()
            _data.cards?.forEachIndexed { index, _ ->
                val view = inflater.inflate(R.layout.home_card_pager_indicator, pagerIndicator, false)
                view.isSelected = cardsPager?.currentItem ?: 0 == index
                pagerIndicator?.addView(view)
            }

            cardsPager?.addOnPageChangeListener(object : ViewPager.OnPageChangeListener {
                override fun onPageScrollStateChanged(state: Int) {
                    (parentFragment as KHomeFragment).refreshingEnabled = state == ViewPager.SCROLL_STATE_IDLE
                }

                override fun onPageScrolled(position: Int, positionOffset: Float, positionOffsetPixels: Int) = Unit

                override fun onPageSelected(position: Int) {
                    for (i in 0..(pagerIndicator?.childCount ?: 0)) {
                        pagerIndicator?.getChildAt(i)?.isSelected =
                                (i == position)
                    }
                }
            })
        }
    }

    inner class CardsAdapter : FragmentStatePagerAdapter(childFragmentManager), Observer<Notification<JsonObject>> {

        private var cards: JsonArray? = null

        override fun getItem(position: Int): Fragment =
                when (cards?.get(position)?.asJsonObject.itemType) {
                    TeambrellaModel.FEED_ITEM_PAY_TO_JOIN -> createCardFragment(PayToJoinCardFragment::class.java, position
                            , cards?.get(position)?.asJsonObject?.itemType ?: 0
                            , tags)

                    TeambrellaModel.FEED_ITEM_UPDATE_PROFILE -> createCardFragment(UpdateProfileCardFragment::class.java, position
                            , cards?.get(position)?.asJsonObject?.itemType ?: 0
                            , tags)

                    else -> createCardFragment(ClaimCardFragment::class.java, position
                            , cards?.get(position)?.asJsonObject?.itemType ?: 0
                            , tags)
                }


        override fun getCount() = cards?.size() ?: 0
        override fun getItemPosition(fragment: Any): Int {
            if (fragment is CardFragment) {
                return if (fragment.type == cards?.get(fragment.position)?.asJsonObject?.itemType) PagerAdapter.POSITION_UNCHANGED else
                    PagerAdapter.POSITION_NONE
            }
            return PagerAdapter.POSITION_NONE
        }

        override fun onChanged(notification: Notification<JsonObject>?) {
            if (notification?.isOnNext == true) {
                cards = notification.value?.data?.cards
                notifyDataSetChanged()
            }
        }
    }
}

private const val EXTRA_POSITION = "position"
private const val EXTRA_TYPE = "type"

private fun <T : CardFragment> createCardFragment(clazz: Class<T>, position: Int, type: Int, tags: Array<String>) = createDataFragment(tags, clazz).apply {
    arguments?.putInt(EXTRA_POSITION, position)
    arguments?.putInt(EXTRA_TYPE, type)
}


abstract class CardFragment : ADataFragment<IMainDataHost>() {
    val position: Int by lazy(LazyThreadSafetyMode.NONE) {
        arguments?.getInt(EXTRA_POSITION, 0) ?: 0
    }
    val type: Int by lazy(LazyThreadSafetyMode.NONE) {
        arguments?.getInt(EXTRA_TYPE, 0) ?: 0
    }
}

class PayToJoinCardFragment : CardFragment() {

    private val titleView: TextView? by ViewHolder(R.id.title)
    private val textView: TextView? by ViewHolder(R.id.text)
    private val subtitle: TextView? by ViewHolder(R.id.subtitle)
    private val icon: ImageView? by ViewHolder(R.id.icon)

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? =
            inflater.inflate(R.layout.home_cards_action, container, false)


    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        view.setOnClickListener { _ ->
            dataHost.showWallet()
        }
    }

    override fun onDataUpdated(notification: Notification<JsonObject>) {
        notification.takeIf { it.isOnNext }?.value?.data?.cards?.get(position)?.asJsonObject?.let {
            titleView?.text = it.chatTitle
            textView?.text = it.text
            subtitle?.text = it.subTitle
            icon?.setImage(imageLoader.getImageUrl(it.smallPhotoOrAvatar), R.dimen.rounded_corners_3dp)
        }
    }
}


class UpdateProfileCardFragment : CardFragment() {

    private val titleView: TextView? by ViewHolder(R.id.title)
    private val textView: TextView? by ViewHolder(R.id.text)
    private val subtitle: TextView? by ViewHolder(R.id.subtitle)
    private val icon: ImageView? by ViewHolder(R.id.icon)

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? =
            inflater.inflate(R.layout.home_cards_action, container, false)


    override fun onDataUpdated(notification: Notification<JsonObject>) {
        notification.takeIf { it.isOnNext }?.value?.data?.cards?.get(position)?.asJsonObject?.let {
            titleView?.text = it.chatTitle
            textView?.text = it.text
            subtitle?.text = it.subTitle
            icon?.setImage(imageLoader.getImageUrl(it.smallPhotoOrAvatar), R.dimen.rounded_corners_3dp)


            view?.setOnClickListener { _ ->
                startActivity(ChatActivity.getTeammateChat(context, dataHost.teamId
                        , it.itemUserId
                        , it.itemUserName
                        , null
                        , it.topicId
                        , dataHost.teamAccessLevel))
            }
        }
    }
}


class ClaimCardFragment : CardFragment() {

    private val icon: ImageView? by ViewHolder(R.id.icon)
    private val teammatePicture: ImageView? by ViewHolder(R.id.teammate_picture)
    private val unread: TextView? by ViewHolder(R.id.unread)
    private val amountWidget: TextView? by ViewHolder(R.id.amount_widget)
    private val teamVote: TextView? by ViewHolder(R.id.team_vote)
    private val title: TextView? by ViewHolder(R.id.title)
    private val subtitle: TextView? by ViewHolder(R.id.subtitle)
    private val leftTile: TextView? by ViewHolder(R.id.left_title)
    private val votingLabel: TextView? by ViewHolder(R.id.voting_label)
    private val message: TextView? by ViewHolder(R.id.message_text)

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? =
            inflater.inflate(R.layout.home_card_claim, container, false)


    override fun onDataUpdated(notification: Notification<JsonObject>) {
        val card = notification.takeIf { it.isOnNext }?.value?.data?.cards?.get(position)?.asJsonObject
        card?.let { _card ->
            when (_card.itemType) {
                TeambrellaModel.FEED_ITEM_TEAMMATE -> {
                    leftTile?.setText(R.string.coverage)
                    icon?.setAvatar(imageLoader.getImageUrl(_card.smallPhotoOrAvatar))
                    title?.text = _card.itemUserName
                    teamVote?.text = String.format(Locale.US, "%.1f", _card.teamVote ?: 0f)
                    subtitle?.text = getString(R.string.object_format_string, _card.modelOrName, _card.year)
                    teammatePicture?.visibility = View.GONE
                    subtitle?.layoutParams = (subtitle?.layoutParams as ConstraintLayout.LayoutParams).apply {
                        marginStart = 0
                    }
                }
                TeambrellaModel.FEED_ITEM_CLAIM -> {
                    leftTile?.setText(R.string.claimed)
                    icon?.setImage(imageLoader.getImageUrl(_card.smallPhotoOrAvatar), R.dimen.rounded_corners_3dp)
                    title?.text = _card.modelOrName
                    teamVote?.text = Html.fromHtml(getString(R.string.home_team_vote_format_string, Math.round((_card.teamVote
                            ?: 0f) * 100)))
                    subtitle?.text = _card.itemUserName
                    teammatePicture?.visibility = View.VISIBLE
                    teammatePicture?.setAvatar(imageLoader.getImageUrl(_card.itemUserAvatar))
                    subtitle?.layoutParams = (subtitle?.layoutParams as ConstraintLayout.LayoutParams).apply {
                        marginStart = resources.getDimensionPixelOffset(R.dimen.margin_4)
                    }
                }
            }

            message?.text = Html.fromHtml(_card.text)
            message?.post { message?.maxLines = if ((message?.length() ?: 0) > 64) 2 else 1 }


            unread?.setUnreadCount(_card.unreadCount ?: 0)

            AmountCurrencyUtil.setAmount(amountWidget, _card.amount ?: 0f, dataHost.currency)

            votingLabel?.visibility = if (_card.isVoting == true) View.VISIBLE else View.GONE

            view?.setOnClickListener {
                when (_card.itemType) {
                    TeambrellaModel.FEED_ITEM_CLAIM -> {
                        startActivity(ChatActivity.getClaimChat(context
                                , dataHost.teamId
                                , _card.itemIdInt ?: 0
                                , _card.modelOrName
                                , _card.smallPhotoOrAvatar
                                , _card.topicId
                                , dataHost.teamAccessLevel
                                , _card.itemDate))
                    }
                    else -> {
                        startActivity(ChatActivity.getTeammateChat(context, dataHost.teamId
                                , _card.itemUserId
                                , _card.itemUserName
                                , _card.smallPhotoOrAvatar
                                , _card.topicId
                                , dataHost.teamAccessLevel))
                    }
                }
            }
        }
    }
}